types.go

```diff
@@ -34,101 +38,51 @@ func PHPString(s string, persistent bool) unsafe.Pointer {
 	return unsafe.Pointer(zendStr)
 }
 
-// PHPKeyType represents the type of PHP hashmap key
-type PHPKeyType int
-
-const (
-	PHPIntKey PHPKeyType = iota
-	PHPStringKey
-)
-
-type PHPKey struct {
-	Type PHPKeyType
-	Str  string
-	Int  int64
-}
-
-// Array represents a PHP array with ordered key-value pairs
-type Array struct {
-	keys   []PHPKey
-	values []interface{}
+// AssociativeArray represents a PHP array with ordered key-value pairs
+type AssociativeArray struct {
+	Map   map[string]interface{}
+	Order []string
 }
 
-// SetInt sets a value with an integer key
-func (arr *Array) SetInt(key int64, value interface{}) {
-	arr.keys = append(arr.keys, PHPKey{Type: PHPIntKey, Int: key})
-	arr.values = append(arr.values, value)
-}
-
-// SetString sets a value with a string key
-func (arr *Array) SetString(key string, value interface{}) {
-	arr.keys = append(arr.keys, PHPKey{Type: PHPStringKey, Str: key})
-	arr.values = append(arr.values, value)
-}
-
-// Append adds a value to the end of the array with the next available integer key
-func (arr *Array) Append(value interface{}) {
-	nextKey := arr.getNextIntKey()
-	arr.SetInt(nextKey, value)
-}
-
-// getNextIntKey finds the next available integer key
-func (arr *Array) getNextIntKey() int64 {
-	maxKey := int64(-1)
-	for _, key := range arr.keys {
-		if key.Type == PHPIntKey && key.Int > maxKey {
-			maxKey = key.Int
-		}
-	}
-
-	return maxKey + 1
-}
-
-// Len returns the number of elements in the array
-func (arr *Array) Len() uint32 {
-	return uint32(len(arr.keys))
-}
-
-// At returns the key and value at the given index
-func (arr *Array) At(index uint32) (PHPKey, interface{}) {
-	if index >= uint32(len(arr.keys)) {
-		return PHPKey{}, nil
-	}
-	return arr.keys[index], arr.values[index]
-}
-
-// EXPERIMENTAL: GoArray converts a zend_array to a Go Array
-func GoArray(arr unsafe.Pointer) *Array {
-	result := &Array{
-		keys:   make([]PHPKey, 0),
-		values: make([]interface{}, 0),
-	}
+// PackedArray represents a 'packed' PHP array as an indexed list
+type PackedArray = []interface{}
 
+// EXPERIMENTAL: GoAssociativeArray converts a zend_array to a Go AssociativeArray
+func GoAssociativeArray(arr unsafe.Pointer, ordered bool) (AssociativeArray, error) {
 	if arr == nil {
-		return result
+		return AssociativeArray{}, errors.New("GoAssociativeArray received a nil pointer")
 	}
 
 	zval := (*C.zval)(arr)
 	hashTable := (*C.HashTable)(castZval(zval, C.IS_ARRAY))
 
 	if hashTable == nil {
-		return result
+		return AssociativeArray{}, errors.New("GoAssociativeArray received a zval that was not a HashTable")
```

Comment:
I wonder if this function shouldn't panic instead of returning errors, as these two errors are clearly programming errors that should be fixed in the code itself.

Reply:
IMHO we should allow "bad" PHP code but fail as early and hardly as possible for bad Go or C code.